qiskit_ode/signals/signals.py

```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
+    # convert back
+    if isinstance(sig1, DiscreteSignal) and isinstance(sig2, DiscreteSignal):
+        if (
+            sig1.dt == sig2.dt
+            and sig1.start_time == sig2.start_time
+            and sig1.duration == sig2.duration
+        ):
+            sig_sum = DiscreteSignalSum.from_SignalSum(
+                sig_sum, dt=sig2.dt, start_time=sig2.start_time, n_samples=sig2.duration
+            )
+
+    return sig_sum
+
+
+def signal_multiply(sig1: Signal, sig2: Signal) -> SignalSum:
+    r"""Multiply two ``Signal``s. For a pair of elementary (non-``SignalSum``) ``Signal``s,
+    expands the product of two signals into a ``SignalSum`` via the formula:
+
+    .. math::
+        Re[f(t)e^{i(2 \pi \nu t + \phi)}] \times Re[g(t)e^{i(2 \pi \omega t + \psi)}]
+         = Re[\frac{1}{2} f(t)g(t)e^{i(2\pi (\omega + \nu)t + (\phi + \psi))} ]
+          + Re[\frac{1}{2} f(t)\overline{g(t)}e^{i(2\pi (\omega - \nu)t + (\phi - \psi))} ]
+
+    If either (or both) of ``sig1`` or ``sig2`` are ``SignalSum``s, the multiplication is
+    distributed over addition.
+    """
+
+    # convert to SignalSum instances
+    try:
+        wrapped_sig1 = to_SignalSum(sig1)
+        wrapped_sig2 = to_SignalSum(sig2)
+    except QiskitError as e:
+        raise QiskitError("Only a number or a Signal instance can multiply a Signal.") from e
+
+    # initialize to empty sum
+    product = SignalSum()
+
+    # loop through every pair of components and multiply
+    for comp1, comp2 in itertools.product(wrapped_sig1.components, wrapped_sig2.components):
+        product += base_signal_multiply(comp1, comp2)
+
+    # if they were originally DiscreteSignalSum objects with compatible structure,
+    # convert back
+    if isinstance(sig1, DiscreteSignalSum) and isinstance(sig2, DiscreteSignalSum):
+        if (
+            sig1.dt == sig2.dt
+            and sig1.start_time == sig2.start_time
+            and sig1.duration == sig2.duration
+        ):
+            product = DiscreteSignalSum.from_SignalSum(
+                product, dt=sig1.dt, start_time=sig1.start_time, n_samples=sig1.duration
+            )
+
+    return product
+
+
+def base_signal_multiply(sig1: Signal, sig2: Signal) -> Signal:
+    r"""Utility function for multiplying two elementary (non ``SignalSum``) signals.
+    This function assumes ``sig1`` and ``sig2`` are legitimate instances of ``Signal``
+    subclasses.
+
+    Special cases:
+
+        - Multiplication of two ``Constant``s returns a ``Constant``.
+        - Multiplication of a ``Constant`` and a ``DiscreteSignal`` returns a ``DiscreteSignal``.
+        - If two ``DiscreteSignal``s have compatible parameters, the resulting signals are
+        ``DiscreteSignal``, with the multiplication being implemented by array multiplication of
+        the samples.
+        - Lastly, if no special rules apply, the two ``Signal``s are multiplied generically via
+        multiplication of the envelopes as functions.
```

Comment:
What happens to carrier phase and frequency?

Reply:
Updated comment to explain.